lua/plainline/core.lua

```diff
@@ -64,28 +70,98 @@ end
+    return
+  end
+  -- Means current window
+  local window_id = 0
+  -- Give up if window is floating. This is done because
+  -- floating windows don't deal well with winbars.
+  if vim.api.nvim_win_get_config(window_id).relative ~= '' then
+    return
+  end
+
+  if mode == 'on' then
+    vim.wo.winbar = "%{%v:lua.require'plainline.core'.winbar_on()%}"
+  elseif mode == 'off' then
+    vim.wo.winbar = "%{%v:lua.require'plainline.core'.winbar_off()%}"
+  else
+    error(string.format('Invalid winbar mode "%s"', mode))
+  end
+end
+
+local function setup_locals(config)
 
   -- Core autocommands to get plainline running
-  local plainline = vim.api.nvim_create_augroup("plainline", {})
+  local plainline = this.plainline_group
   vim.api.nvim_create_autocmd({ "WinEnter", "BufEnter" }, {
-      command = "setl statusline=%{%v:lua.require'plainline.core'.on()%}",
+    command = "setl statusline=%{%v:lua.require'plainline.core'.on()%}",
+    group = plainline,
+  })
+  vim.api.nvim_create_autocmd({ "WinLeave", "BufLeave" }, {
+    command = "setl statusline=%{%v:lua.require'plainline.core'.off()%}",
+    group = plainline,
+  })
+
+  -- Setup winbar too
+  if config.winbar then
+    vim.api.nvim_create_autocmd({ "WinEnter", "BufEnter" }, {
+      callback = function() enable_winbar('on') end,
       group = plainline,
     })
-  vim.api.nvim_create_autocmd({ "WinLeave", "BufLeave" }, {
-      command = "setl statusline=%{%v:lua.require'plainline.core'.off()%}",
+    vim.api.nvim_create_autocmd({ "WinLeave", "BufLeave" }, {
+      callback = function() enable_winbar('off') end,
       group = plainline,
     })
+  end
+end
+
+local function setup_globals(config)
+  vim.go.statusline = "%{%v:lua.require'plainline.core'.on()%}"
+
+  if config.winbar then
+    -- Unfortunately, this still needs to be run using autocommands
+    vim.api.nvim_create_autocmd({ "WinEnter", "BufEnter" }, {
+      callback = function() enable_winbar('on') end,
+      group = this.plainline_group,
+    })
+  end
+end
+
+-- Enables plainline, using the functions in plainline.core to set up the
+-- appropriate autocommands for the thing to work properly
+function this.enable(config)
+  local on = get_ptable(config.sections)
+  local off = get_ptable(config.inactive_sections)
+
+  this.plainline_group = vim.api.nvim_create_augroup("plainline", {})
+
+  -- Statusline functions for active and inactive states, respectively
+  this.on  = function() return mkstatus(on, config) end
+  this.off = function() return mkstatus(off, config) end
+
+  if config.winbar then
+    local winbar_on = get_ptable(config.winbar.sections or {})
+    local winbar_off = get_ptable(config.winbar.inactive_sections or {})
+
+    -- Winbar functions for active and inactive states, respectively
+    this.winbar_on  = function() return mkstatus(winbar_on, config) end
+    this.winbar_off = function() return mkstatus(winbar_off, config) end
+  end
+
+  if not config.global then
```

Comment:
I don't think we need to introduce a new key in the configuration for this. If the user has `vim.o.laststatus` set to 3, then what they want is the global setup. If it's not 3, the global setup doesn't really make sense. So it would be better to check `laststatus` here, I think.

Reply:
I think this is fair